src/util/future.rs

```diff
@@ -29,3 +31,62 @@ where
     };
     (callback, future)
 }
+
+/// A shortcut for `f1.and_then(|()| f2.flatten())`. Note that
+/// the expression is just a simplified version as f2's Error
+/// type may not be easy handled via combinators.
+pub struct AndThenWith<F1, F2>
```

Comment:
where to use this?

Reply:
In `kv.rs`, where the result of API call needs to be handled the same way as the value returned by the future.